File: client/common/connectionHandler.go
package common

import (
	"net"
)

type ConnectionHandler struct {
	conn net.Conn
}

func NewConnectionHandler(conn net.Conn) *ConnectionHandler {
	ConnectionHandler := &ConnectionHandler{
		conn: conn,
	}
	return ConnectionHandler
}

func (mc *ConnectionHandler) send(data []byte) error {
	sizeToSend := len(data)

	totalBytesSent := 0
	for totalBytesSent < sizeToSend {
		bytesSent, err := mc.conn.Write(data[totalBytesSent:])

		if err != nil {
			return err
		}

		totalBytesSent += bytesSent
	}

	return nil
}

func (mc *ConnectionHandler) recv(sizeToRecv int) ([]byte, error) {
	log.Debugf("Se desea recibir %v", sizeToRecv)
	data := make([]byte, sizeToRecv)

	totalBytesRecv := 0
	for totalBytesRecv < sizeToRecv {
		bytesRecv, err := mc.conn.Read(data[totalBytesRecv:])

		log.Debugf("Se recibe %v | %v", bytesRecv, err)

		if err != nil {
			return nil, err
		}

		totalBytesRecv += bytesRecv
	}

	return data, nil
}

func (c *ConnectionHandler) close() {
	if c.conn != nil {
		c.conn.Close()
	}
}

File: server/common/connectionHandler.py
import socket
import logging

from common.exceptions import ClientConnectionClosedException

class ConnectionHandler:
    def __init__(self, client_sock: socket):
        self._socket = client_sock

    def send(self, data):
        size_to_send = len(data)
        total_bytes_sent = 0
        
        while total_bytes_sent < size_to_send:
            bytes_sent = self._socket.send(data[total_bytes_sent:])

            if bytes_sent == 0:
                raise ClientConnectionClosedException()

            total_bytes_sent += bytes_sent

    def recv(self, size_to_recv):
        total_bytes_recv = 0
        data = b''
        
        while total_bytes_recv < size_to_recv:
            data_chunk = self._socket.recv(size_to_recv - total_bytes_recv)
            
            if data_chunk == b'':
                raise ClientConnectionClosedException()
            
            total_bytes_recv += len(data_chunk)
            data += data_chunk
            
        return data
    
    def close(self):
        self._socket.close()
        